src/main/kotlin/org/rust/lang/core/resolve/scope/RustResolveScope.kt

```diff
@@ -1,9 +1,258 @@
+    val module = module ?: return emptySequence()
+    // Rust injects implicit `extern crate std` in every crate root module unless it is
+    // a `#![no_std]` crate, in which case `extern crate core` is injected.
+    // The stdlib lib itself is `#![no_std]`.
+    // We inject both crates for simplicity for now.
+    val injectedCrates = if (isCrateRoot) {
+        sequenceOf(AutoInjectedCrates.std, AutoInjectedCrates.core).mapNotNull { crateName ->
+            ScopeEntryImpl.lazy(crateName) {
+                module.project.getPsiFor(module.cargoProject?.findExternCrateRootByName(crateName))?.rustMod
+            }
+        }
+    } else emptySequence<RustResolveScope.Entry>()
+
+    // Rust injects implicit `use std::prelude::v1::*` into every module.
+    val preludeSymbols = if (context.inPrelude) emptySequence() else module.preludeDeclarations(context)
+
+    return injectedCrates + preludeSymbols
+}
+
+private fun Module.preludeDeclarations(context: RustResolveScope.Context): Sequence<RustResolveScope.Entry> =
+    preludeModule?.rustMod?.declarations(context.copy(inPrelude = true)) ?: emptySequence()
+
+private fun RustUseItemElement.importedEntries(context: RustResolveScope.Context): Sequence<RustResolveScope.Entry> {
+    if (isStarImport) {
+        if (this in context.visitedStarImports) return emptySequence()
+        // Recursively step into `use foo::*`
+        val pathPart = path ?: return emptySequence()
+        val mod = pathPart.reference.resolve() as? RustResolveScope ?: return emptySequence()
+        return mod.declarations(context.copy(visitedStarImports = context.visitedStarImports + this))
+    }
+
+    val globList = useGlobList
+    if (globList == null) {
+        val path = path ?: return emptySequence()
+        // use foo::bar [as baz];
+        val entry = ScopeEntryImpl.lazy(alias?.name ?: path.referenceName) { path.reference.resolve() }
+        return listOfNotNull(entry).asSequence()
+    }
+
+    return globList.useGlobList.asSequence().mapNotNull { glob ->
+        val name = listOfNotNull(
+            glob.alias?.name, // {foo as bar};
+            glob.self?.let { path?.referenceName }, // {self}
+            glob.referenceName // {foo}
+        ).firstOrNull()
+
+        ScopeEntryImpl.lazy(name) { glob.reference.resolve() }
+    }
+}
+
+/**
+ *  Let declarations visible at the `element` according to Rust scoping rules.
+ *  More recent declarations come first.
+ *
+ *  Example:
+ *
+ *    ```
+ *    {
+ *        let x = 92; // visible
+ *        let x = x;  // not visible
+ *                ^ element
+ *        let x = 62; // not visible
+ *    }
+ *    ```
+ */
+private fun RustBlockElement.letDeclarationsVisibleAt(element: RustCompositeElement): Sequence<RustLetDeclElement> =
+    letDeclarations
+        .dropWhile { PsiUtilCore.compareElementsByPosition(element, it) < 0 }
+        // Drops at most one element
```

Comment:
Yeah. In the inner block there is only one `let`, and it will be dropped. In the outer block, `let x = 92` won't be dropped. This function looks only at one block at a time, and for both blocks it drops exactly one `let.`

However your example indeed demonstrates a pathological behavior: for a series of `n` nested lets, the total cost of `isAncestor` checks is `O(n^2)`. I hope this won't create any troubles because even two level deep nesting of `let`s is pretty rare.